Sources/Datadog/Tracing/Integrations/TracingURLSessionHandler.swift

```diff
@@ -0,0 +1,182 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import Foundation
+import DatadogInternal
+
+internal struct TracingURLSessionHandler: DatadogURLSessionHandler {
+    /// Integration with Core Context.
+    let contextReceiver: ContextMessageReceiver
+    /// Tracing sampler used to sample traces generated by the SDK.
+    let tracingSampler: Sampler
+    /// First party hosts defined by the user.
+    let firstPartyHosts: FirstPartyHosts
+
+    weak var tracer: DatadogTracer?
+
+    init(
+        tracer: DatadogTracer,
+        contextReceiver: ContextMessageReceiver,
+        tracingSampler: Sampler,
+        firstPartyHosts: FirstPartyHosts
+    ) {
+        self.tracer = tracer
+        self.contextReceiver = contextReceiver
+        self.tracingSampler = tracingSampler
+        self.firstPartyHosts = firstPartyHosts
+    }
+
+    init(
+        tracer: DatadogTracer,
+        contextReceiver: ContextMessageReceiver,
+        distributedTracingConfiguration: DatadogTracer.DistributedTracingConfiguration
+    ) {
+        self.init(
+            tracer: tracer,
+            contextReceiver: contextReceiver,
+            tracingSampler: .init(samplingRate: distributedTracingConfiguration.tracingSamplingRate),
+            firstPartyHosts: .init(distributedTracingConfiguration.firstPartyHosts)
+        )
+    }
+
+    func modify(request: URLRequest, headerTypes: Set<DatadogInternal.TracingHeaderType>) -> URLRequest {
+        guard let tracer = tracer else {
+            return request
+        }
+
+        let spanContext = tracer.createSpanContext()
+
+        var request = request
+        headerTypes.forEach {
+            let writer: TracePropagationHeadersWriter
+            switch $0 {
+            case .datadog:
+                writer = HTTPHeadersWriter(sampler: tracingSampler)
+            case .b3:
+                writer = OTelHTTPHeadersWriter(
+                    sampler: tracingSampler,
+                    injectEncoding: .single
+                )
+            case .b3multi:
+                writer = OTelHTTPHeadersWriter(
+                    sampler: tracingSampler,
+                    injectEncoding: .multiple
+                )
+            case .tracecontext:
+                writer = W3CHTTPHeadersWriter(sampler: tracingSampler)
+            }
+
+            writer.write(
+                traceID: spanContext.traceID,
+                spanID: spanContext.spanID,
+                parentSpanID: spanContext.parentSpanID
+            )
+
+            writer.traceHeaderFields.forEach { field, value in
+                request.setValue(value, forHTTPHeaderField: field)
+            }
+        }
+
+        return request
+    }
+
+    func isInternal(request: URLRequest) -> Bool {
+        return false
+    }
```

Comment:
I thought we will remove it - we validated that "internal URLs filtering" is no longer required. Is it a leftover?